### PR TITLE
planner: reset recorded cte plan in plan encoder

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25174

Problem Summary:

planner encoder does not reset the recorded cte plan lead data race

### Release note 

- reset recorded cte plan in plan encoder

### CODE REVIEW

**Review by guo-shaoge — APPROVED** [review]

File: planner/core/encode.go [hunk 1]
```diff
@@ -60,6 +60,7 @@ func EncodePlan(p Plan) string {
 func (pn *planEncoder) encodePlanTree(p Plan) string {
 	pn.encodedPlans = make(map[int]bool)
 	pn.buf.Reset()
+	pn.ctes = pn.ctes[:0]
 	pn.encodePlan(p, true, kv.TiKV, 0)
 	pn.encodeCTEPlan()
```
Comment (wjhuang2016) [inline planner/core/encode.go:65]: It won't work after `pn.ctes = pn.ctes[:0]`

**Review by wjhuang2016 — APPROVED** [review]

File: planner/core/encode.go (hunk 1 above)
Comment (winoros) [inline planner/core/encode.go:65]: We first clear the cte. Then the `encodePlan` would fill the `ctes` and generate outputs by `encodeCTEPlan`?

Comment (xiongjiwei) [reply planner/core/encode.go:65]: right

**Review by winoros — APPROVED** [review]

**Pull request merged**